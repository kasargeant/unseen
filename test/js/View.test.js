/**
 * @file Unseen.test.js
 * @description Unit tests for the Unseen ES2015 class library.
 * @author Kyle Alexis Sargeant <kasargeant@gmail.com> {@link https://github.com/kasargeant https://github.com/kasargeant}.
 * @copyright Kyle Alexis Sargeant 2017
 * @license See LICENSE file included in this distribution.
 */

"use strict";

// Environment
const IS_CI = process.env.CI;
const IS_TRAVIS = process.env.TRAVIS;

// Imports
const fs = require("fs");
const path = require("path");
// const sinon = require("sinon");

// Unit
const Unseen = require("../../src/index");

// Constants
let workingDirectory = process.cwd();
if(IS_TRAVIS) {workingDirectory = process.env.TRAVIS_BUILD_DIR;} // Usually: "/home/travis/build/kasargeant/warhorse"

// Setup
let rawData = require("../data/processed_sample.json");

// const schema = {"id": 0, "idn": "unnamed", "class": "unknown", "type": "unknown", "name": "Unnamed"};

// MODEL
let myModelInstance = new Unseen.Model({"id": 123, "idn": "015695954", "type": "test", "name": "Test Street"});

// MODEL COLLECTION
let myModelCollection = new Unseen.ModelCollection(rawData);

// VIEW
class MyView extends Unseen.View {

    initialize() {
        this.baseModel = myModelInstance;
        this.id = "my-item";
        this.tag = "div";
        this.classList = ["card"];
    }

    events() {
        return {
            "#button-delete": ["click", "deleteAction"]
        };
    }

    template(model, idx) {

        return `
            <div class="card-header">
                <h4 class="card-title">${model.id}</h4>
                <h6 class="card-subtitle">${model.name}</h6>
            </div>
            <div class="card-body">
                ${model.id}: ${model.type} - ${model.name}
            </div>
            <div class="card-footer">
                <button id="button-delete" class="btn btn-primary">Delete</button>
            </div>
        `;
    }

    deleteAction(evt) {
        console.log(`deleteAction for ${this._id} called by ${JSON.stringify(evt)}.`);
        this.destroy();
    }
}
let myViewInstance = new MyView(myModelInstance);


// VIEW COLLECTION
class MyViewCollection extends Unseen.ViewCollection {
    initialize() {
        this.baseClass = MyView;
        this.id = "my-list";
        this.tag = "div";
        this.classList = ["container"];
    }
}
let myViewCollectionInstance = new MyViewCollection(myModelCollection);
myViewCollectionInstance.fetch(false);


console.log(`Testing with ${myModelCollection.length} records.`);

// Tests
describe("Class: View", function() {

    describe("Standard sanity check", function() {
        it("contains spec with an positive expectation", function() {
            expect(true).toBe(true);
        });
        it("contains spec with a negative expectation", function() {
            expect(!true).toBe(false);
        });
    });

    describe("View", function() {

        it("should return render text output.", function() {
            let markupResult = {html: ""};
            myViewInstance._renderMarkup(false, markupResult);
            expect(markupResult.html).not.toBe("");
            expect(markupResult.html).toMatchSnapshot();
        });

    });
    describe("ViewCollection", function() {

        it("should return render text output.", function() {
            let markupResult = {html: ""};
            myViewCollectionInstance._renderMarkup(false, markupResult);
            expect(markupResult.html).not.toBe("");
            expect(markupResult.html).toMatchSnapshot();
        });

    });

});
