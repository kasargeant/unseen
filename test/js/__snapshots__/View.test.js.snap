// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class: View View should return render text output. 1`] = `
"<div id=\\"my-item-0\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">123</h4>
                <h6 class=\\"card-subtitle\\">Test Street</h6>
            </div>
            <div class=\\"card-body\\">
                123: test - Test Street
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-0\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div>"
`;

exports[`Class: View ViewCollection should return render text output. 1`] = `
"<div id=\\"my-list-0\\" class=\\"my-list container\\"><div id=\\"my-item-0\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100011508</h4>
                <h6 class=\\"card-subtitle\\">St. Mary's Gate</h6>
            </div>
            <div class=\\"card-body\\">
                100011508: residential - St. Mary's Gate
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-0\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-1\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100013440</h4>
                <h6 class=\\"card-subtitle\\">La Perla</h6>
            </div>
            <div class=\\"card-body\\">
                100013440: restaurant - La Perla
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-1\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-2\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100013447</h4>
                <h6 class=\\"card-subtitle\\">The Lady</h6>
            </div>
            <div class=\\"card-body\\">
                100013447: unknown - The Lady
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-2\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-3\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015712</h4>
                <h6 class=\\"card-subtitle\\">Carlisle Lane</h6>
            </div>
            <div class=\\"card-body\\">
                100015712: unclassified - Carlisle Lane
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-3\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-4\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015713</h4>
                <h6 class=\\"card-subtitle\\">Virgil Street</h6>
            </div>
            <div class=\\"card-body\\">
                100015713: unclassified - Virgil Street
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-4\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-5\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015714</h4>
                <h6 class=\\"card-subtitle\\">Virgil Street</h6>
            </div>
            <div class=\\"card-body\\">
                100015714: unclassified - Virgil Street
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-5\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-6\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015715</h4>
                <h6 class=\\"card-subtitle\\">Carlisle Lane</h6>
            </div>
            <div class=\\"card-body\\">
                100015715: unclassified - Carlisle Lane
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-6\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-7\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015716</h4>
                <h6 class=\\"card-subtitle\\">Centaur Street</h6>
            </div>
            <div class=\\"card-body\\">
                100015716: unclassified - Centaur Street
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-7\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-8\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015720</h4>
                <h6 class=\\"card-subtitle\\">Upper Marsh</h6>
            </div>
            <div class=\\"card-body\\">
                100015720: unclassified - Upper Marsh
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-8\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-9\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015721</h4>
                <h6 class=\\"card-subtitle\\">Upper Marsh</h6>
            </div>
            <div class=\\"card-body\\">
                100015721: unclassified - Upper Marsh
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-9\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-10\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100015723</h4>
                <h6 class=\\"card-subtitle\\">Station Approach</h6>
            </div>
            <div class=\\"card-body\\">
                100015723: unclassified - Station Approach
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-10\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-11\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100026128</h4>
                <h6 class=\\"card-subtitle\\">Sterling Way</h6>
            </div>
            <div class=\\"card-body\\">
                100026128: trunk - Sterling Way
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-11\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-12\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100026129</h4>
                <h6 class=\\"card-subtitle\\">North Circular Road</h6>
            </div>
            <div class=\\"card-body\\">
                100026129: trunk - North Circular Road
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-12\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-13\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100026130</h4>
                <h6 class=\\"card-subtitle\\">Sterling Way</h6>
            </div>
            <div class=\\"card-body\\">
                100026130: trunk - Sterling Way
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-13\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-14\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">100026131</h4>
                <h6 class=\\"card-subtitle\\">North Circular Road</h6>
            </div>
            <div class=\\"card-body\\">
                100026131: trunk - North Circular Road
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-14\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-15\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">1000463949</h4>
                <h6 class=\\"card-subtitle\\">Empire</h6>
            </div>
            <div class=\\"card-body\\">
                1000463949: supermarket - Empire
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-15\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div><div id=\\"my-item-16\\" class=\\"my-item card\\">
            <div class=\\"card-header\\">
                <h4 class=\\"card-title\\">1000467325</h4>
                <h6 class=\\"card-subtitle\\">Trusha (indian cuisine)</h6>
            </div>
            <div class=\\"card-body\\">
                1000467325: restaurant - Trusha (indian cuisine)
            </div>
            <div class=\\"card-footer\\">
                <button id=\\"button-delete-16\\" class=\\"btn btn-primary\\">Delete</button>
            </div>
        </div></div>"
`;
